### ISSUE TITLE
It seems to be impossible to safely stop a Event Recorder without invoking a race condition

### ISSUE THREAD

**Issue by DirectXMan12** [kubernetes/kubernetes#94906]
The standard event recorder, when recording an event, calls the underlying Broadcaster in a goroutine, so that even recording is non-blocking.  Things like leader election emit events over the course of their lifecycle via this broadcaster.

Consider the following scenario:

* GR1: spawn leader election --> `go leaderElection.Start(leStop)` --> GR2
* GR2: Leader election: emit event "Lease Acquired" --> `go recorder.Broadcaster.Action(leaseAcquiredEvent)` --> spawn GR3 (https://github.com/kubernetes/client-go/blob/master/tools/record/event.go#L344)
* GR1: Shutdown leader election --> `close(leStop); waitForLEToStop()` --> blocks until GR2 is done
* GR1: Shutdown broadcaster --> `close(broadcaster.incoming)`  (https://github.com/kubernetes/apimachinery/blame/master/pkg/watch/mux.go#L210) 
* GR3: `broadcaster.Action(...)` --> `broadcaster.incoming <- Event{...}` --> send on closed channel (https://github.com/kubernetes/apimachinery/blame/master/pkg/watch/mux.go#L198)

Practically, we've observed this behavior popping up occaisionally in tests in controller-runtime: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_controller-runtime/1155/pull-controller-runtime-test-master/1303831291082313728#1:build-log.txt%3A516

**Comment by DirectXMan12** [comment]
<details>

<summary>relevant race detector output/stack traces</summary>

```
==================
WARNING: DATA RACE
Write at 0x00c0001762b0 by goroutine 171:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:334 +0x0
  k8s.io/client-go/tools/record.(*eventBroadcasterImpl).Shutdown()
      /home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.0/pkg/watch/mux.go:199 +0x68
  sigs.k8s.io/controller-runtime/pkg/internal/recorder.(*Provider).Stop.func1()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder.go:73 +0xa0
Previous read at 0x00c0001762b0 by goroutine 168:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:142 +0x0
  k8s.io/client-go/tools/record.(*recorderImpl).generateEvent.func1()
      /home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.0/pkg/watch/mux.go:189 +0x11b
Goroutine 171 (running) created at:
  sigs.k8s.io/controller-runtime/pkg/internal/recorder.(*Provider).Stop()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder.go:66 +0x8c
  sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).engageStopProcedure()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/manager/internal.go:563 +0x38d
  sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).Start.func1()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/manager/internal.go:468 +0x49
  sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).Start()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/manager/internal.go:518 +0x339
  sigs.k8s.io/controller-runtime/pkg/manager.glob..func3.3.8.3.6()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/manager/manager_test.go:273 +0x22b
Goroutine 168 (finished) created at:
  k8s.io/client-go/tools/record.(*recorderImpl).generateEvent()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/record/event.go:341 +0x42a
  k8s.io/client-go/tools/record.(*recorderImpl).Event()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/record/event.go:349 +0xcf
  k8s.io/client-go/tools/record.(*recorderImpl).Eventf()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/record/event.go:353 +0xd7
  sigs.k8s.io/controller-runtime/pkg/internal/recorder.(*lazyRecorder).Eventf()
      /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder.go:151 +0xf3
  k8s.io/client-go/tools/leaderelection/resourcelock.(*LeaseLock).RecordEvent()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/leaderelection/resourcelock/leaselock.go:85 +0x30b
  k8s.io/client-go/tools/leaderelection/resourcelock.(*MultiLock).RecordEvent()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/leaderelection/resourcelock/multilock.go:88 +0xa5
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).acquire.func1()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/leaderelection/leaderelection.go:251 +0x213
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.0/pkg/util/wait/wait.go:155 +0x6f
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.0/pkg/util/wait/wait.go:156 +0xb3
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.0/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).acquire()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/leaderelection/leaderelection.go:244 +0x2b2
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run()
      /home/prow/go/pkg/mod/k8s.io/client-go@v0.19.0/tools/leaderelection/leaderelection.go:203 +0xee
==================
```

</details>

**Comment by fedebongio** [comment]
@DirectXMan12 were you planning on sending a fix, or need our help to send one to fix it?

**Comment by DirectXMan12** [comment]
I can probably try to figure out a fix if you want, but the whole model for how event broadcasters currently work is inverse to how Go suggests you manage channels (only writers should ever close channels, not readers).  I'll need to do a bit of playing around with things.

**Comment by leilajal** [comment]
cc @caesarxuchao

**Comment by caesarxuchao** [comment]
I think this is fixed by #95664.

Please reopen if it's not the case.

**Comment by vincepri** [comment]
@caesarxuchao @fedebongio Would be possible to cherry-pick this fix in 1.20?
